addons/pkg-maven/common/src/main/java/org/commonjava/indy/pkg/maven/content/MetadataInfoWrapper.java

```diff
@@ -0,0 +1,122 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.indy.pkg.maven.content;
+
+import org.apache.maven.artifact.repository.metadata.Metadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+
+import static org.commonjava.indy.pkg.maven.util.MetadataZipUtils.unzip;
+import static org.commonjava.indy.pkg.maven.util.MetadataZipUtils.zip;
+
+/**
+ * Wrapper MetadataInfo around byte arrays which contain gzipped, serialized Metadata / MetadataInfo content
+ * so we can minimize the heap usage.
+ */
+public class MetadataInfoWrapper
+                extends MetadataInfo
+                implements Externalizable
+{
+    private byte[] metadataBytes;
+
+    private byte[] metadataMergeInfoBytes;
+
+    public MetadataInfoWrapper()
+    {
+    }
+
+    @Override
+    public void setMetadata( Metadata metadata )
+    {
+        try
+        {
+            this.metadataBytes = zip( metadata );
+        }
+        catch ( IOException e )
+        {
+            Logger logger = LoggerFactory.getLogger( getClass() );
+            logger.error( "Failed to zip", e );
+        }
+    }
+
+    @Override
+    public void setMetadataMergeInfo( String metadataMergeInfo )
+    {
+        this.metadataMergeInfoBytes = metadataMergeInfo.getBytes();
+    }
+
+    @Override
+    public Metadata getMetadata()
+    {
+        try
+        {
+            return unzip( metadataBytes );
+        }
+        catch ( Exception e )
+        {
+            Logger logger = LoggerFactory.getLogger( getClass() );
+            logger.error( "Failed to unzip", e );
+        }
+        return null;
+    }
+
+    @Override
+    public String getMetadataMergeInfo()
+    {
+        try
+        {
+            return unzip( metadataMergeInfoBytes );
+        }
+        catch ( Exception e )
+        {
+            Logger logger = LoggerFactory.getLogger( getClass() );
+            logger.error( "Failed to unzip", e );
+        }
+        return null;
+    }
+
+    @Override
+    public void writeExternal( ObjectOutput objectOutput ) throws IOException
+    {
+        objectOutput.writeInt( metadataBytes.length );
+        objectOutput.write( metadataBytes );
+        objectOutput.writeInt( metadataMergeInfoBytes.length );
+        objectOutput.write( metadataMergeInfoBytes );
+    }
+
+    @Override
+    public void readExternal( ObjectInput objectInput ) throws IOException, ClassNotFoundException
+    {
+        int len = objectInput.readInt();
+        metadataBytes = new byte[len];
+        objectInput.read( metadataBytes );
+        len = objectInput.readInt();
+        metadataMergeInfoBytes = new byte[len];
+        objectInput.read( metadataMergeInfoBytes );
+    }
+
+    public MetadataInfoWrapper wrap( MetadataInfo metadataInfo )
```

Comment:
Is there a reason not to simply implement this in MetadataInfo directly?

Actually, another approach might be to keep the caches off-heap and then in the serializer for that, do the zipping for the whole MetadataInfo object there.

Reply:
Although we closed it, I think this is a good question. :)
What I thought was to reduce the memory footprint for the MetadataInfo objects whether or not it is in heap or off-heap. If we use MetadataInfo, we will end up using the metadata & info fields (those are what we want to shrink the size for). So I use a wrapper class to wrap it. I make the wrapper extends it so that we don't need to change the caller's code wherever it is used. But in fact, these two are different objects in terms of how they preserve data. 
Or we can remove those fields from MetadataInfo and only keep the set/get methods, and use the zipped bytes to really preserve the content. This is a valid alternative too. But I choose the wrapper approach so the change is less drastic and people can understand it a bit easier.